src/pudl/extract/dbf.py

```diff
@@ -0,0 +1,301 @@
+"""Generalized DBF extractor for FERC data."""
+import csv
+import importlib
+from collections import defaultdict
+from collections.abc import Iterator
+from functools import lru_cache
+from pathlib import Path
+from typing import Any, Protocol
+
+import pandas as pd
+import sqlalchemy as sa
+from dbfread import DBF, FieldParser
+
+from pudl.workspace.datastore import Datastore
+
+
+class TableSchema:
+    """Simple data-wrapper for the fox-pro table schema."""
+
+    def __init__(self, table_name: str):
+        """Creates ne:1w instance of the table schema setting.
+
+        The table name will be set as table_name and table will have no columns.
+        """
+        self.name = table_name
+        self._columns = []
+        self._column_types = {}
+        self._short_name_map = {}  # short_name_map[short_name] -> long_name
+
+    def add_column(
+        self,
+        col_name: str,
+        col_type: sa.types.TypeEngine,
+        short_name: str | None = None,
+    ):
+        """Adds a new column to this table schema."""
+        assert col_name not in self._columns
+        self._columns.append(col_name)
+        self._column_types[col_name] = col_type
+        if short_name is not None:
+            self._short_name_map[short_name] = col_name
+
+    def get_columns(self) -> Iterator[tuple[str, sa.types.TypeEngine]]:
+        """Itereates over the (column_name, column_type) pairs."""
+        for col_name in self._columns:
+            yield (col_name, self._column_types[col_name])
+
+    def get_column_rename_map(self) -> dict[str, str]:
+        """Returns dictionary that maps from short to long column names."""
+        return dict(self._short_name_map)
+
+    def to_sqlite_table(self, sqlite_meta: sa.MetaData) -> sa.Table:
+        """Adds this table schema to SQLAlchemy MetaData object."""
+        table = sa.Table(self.name, sqlite_meta)
+        for col_name, col_type in self.get_columns():
+            table.append_column(sa.Column(col_name, col_type))
+        return table
+
+
+class AbstractFoxProDatastore(Protocol):
+    """This is the interface definition for dealing with fox-pro datastores."""
+
+    def get_dataset(self) -> str:
+        """Returns name of the dataset that this datastore provides access to."""
+        ...
+
+    def get_table_names(self) -> list[str]:
+        """Returns list of all available table names."""
+        ...
+
+    def get_table_schema(self, table_name: str, year: int) -> TableSchema:
```

Comment:
Curious why `get_table_schema` is part of the interface, and `get_db_schema` isn't.

Reply:
`get_table_schema` is used by the `FoxProExtractor`, while `get_db_schema` is only used internally as part of dataframe loading.